rtt_rosparam/include/rtt_rosparam/RosParamDataSource.hpp

```diff
@@ -0,0 +1,221 @@
+ *   produce an executable, this file does not by itself cause the         *
+ *   resulting executable to be covered by the GNU General Public          *
+ *   License.  This exception does not however invalidate any other        *
+ *   reasons why the executable file might be covered by the GNU General   *
+ *   Public License.                                                       *
+ *                                                                         *
+ *   This library is distributed in the hope that it will be useful,       *
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU     *
+ *   Lesser General Public License for more details.                       *
+ *                                                                         *
+ *   You should have received a copy of the GNU General Public             *
+ *   License along with this library; if not, write to the Free Software   *
+ *   Foundation, Inc., 59 Temple Place,                                    *
+ *   Suite 330, Boston, MA  02111-1307  USA                                *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#ifndef __RTT_ROSPARAM__ROSPARAMDATASOURCE_HPP
+#define __RTT_ROSPARAM__ROSPARAMDATASOURCE_HPP
+
+// #include "mystd.hpp"
+#include "rtt/internal/DataSource.hpp"
+#include "rtt/internal/DataSourceTypeInfo.hpp"
+#include "rtt/internal/Reference.hpp"
+
+#include "ros/ros.h"
+#include "ros/param.h"
+#include "ros/names.h"
+
+#include <boost/function.hpp>
+#include <exception>
+
+namespace RTT
+{
+  
+  namespace rosparam {
+
+  /**
+   * A DataSource which checks the ROS parameter server and it
+   * returns the parameter value in its get() method.
+   * It sets the parameter value in its set() method.
+   * The constructor receives the full qualified property name attached
+   * to the DataSource.
+   */
+  template<typename T>
+  class RosParamDataSource
+      : public internal::AssignableDataSource<T>
+  {
+    /**
+     * Stores the full-qualified property name of the ROS parameter
+     */
+    std::string mparam_name_;
+    bool muse_cached_;
+    mutable typename internal::DataSource<T>::value_t mcached_data_;
+
+  public:
+    /**
+     * Use shared_ptr.
+     */
+    ~RosParamDataSource();
+
+    typedef boost::intrusive_ptr< RosParamDataSource<T> > shared_ptr;
+
+    RosParamDataSource(std::string param_name_id, bool use_cached = true);
+
+    typename internal::DataSource<T>::result_t get() const
+    {
+      T return_value;
+      if (muse_cached_) {
+        if (!ros::param::getCached(mparam_name_, return_value)) {
+          // std::cerr << "The parameter " << mparam_name_ << " cannot be fetched." << std::endl;
+          throw(std::range_error("The parameter " + mparam_name_ + " cannot be fetched."));
```

Comment:
```suggestion
          throw std::range_error("The parameter " + mparam_name_ + " cannot be fetched.");
```

`throw` is a C++ keyword, and not a function to call.

Reply:
Why [std::range_error](https://en.cppreference.com/w/cpp/error/range_error)?!? That does not fit in this case. Maybe you meant [std::out_of_range](http://www.cplusplus.com/reference/stdexcept/out_of_range/), which is what [std::map::at(key)](http://www.cplusplus.com/reference/map/map/at/) throws if `key` is not an element in the map. Or just [std::runtime_error](https://en.cppreference.com/w/cpp/error/runtime_error).

But actually a data source should not throw and always return a default value. Only `evaluate()` can fail. So users can call `evaluate()` if they want to know whether the parameter is set, and then get the value with `value()` or `rvalue()`. `get()` should just return a default-constructed value or the previous cached value if `ros::param::get()` fails.

Typically `get()` is just implemented as
```cpp
    typename internal::DataSource<T>::result_t get() const
    {
      (void) evaluate();
      return value();
    }
```
and the actual work of updating `mcached_data_` happens in `evaluate()`.